nixos/modules/services/web-servers/webhook.nix

```
@@ -0,0 +1,91 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.webhook;
+  webhook = pkgs.webhook;
+  user = if (cfg.user == null) then "webhook" else cfg.user;
+  group = if (cfg.group == null) then "webhook" else cfg.group;
+  settingsFormat = pkgs.formats.json {};
+  configFile = settingsFormat.generate "webhook.json" cfg.settings;
+
+in {
+  options.services.webhook = {
+
+    enable = mkEnableOption "the webhook service";
+
+    user = mkOption {
+      default = null;
+      description = "User webhook service runs as. If <literal>null</literal>, it will default to <literal>webhook</literal>";
+      type = with types; nullOr str;
+    };
+
+    group = mkOption {
+      default = null;
+      description = "Group webhook service runs as. If <literal>null</literal>, it will default to <literal>webhook</literal>";
+      type = with types; nullOr str;
+    };
+
+    settings = mkOption {
+      description = "Webhook service settings.";
+      type = settingsFormat.type;
+      example = literalExpression ''
+        [{
+          id = "redeploy-webhook";
+          execute-command = "/var/scripts/redeploy.sh";
+          command-working-directory = "/var/webhook";
+        }]
+      '';
+    };
+
+    path = mkOption {
+      default = [];
+      description = "Which packages the service can access";
+      example = literalExpression ''
+        with pkgs; [ nixFlakes ]
+      '';
+      type = with types; listOf package;
+    };
+
+    extraOptions = mkOption {
+      default = [ "-verbose" ];
+      description = "Which flags to append to the command";
+      example = literalExpression ''
+        [ "-debug" "-port 12345" ]
+      '';
+      type = with types; listOf str;
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.services.webhook = {
+      description = "Webhook service";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      path = cfg.path;
+      serviceConfig = {
+        Type = "simple";
+        User = user;
+        Group = group;
+        WorkingDirectory = "${webhook}";
+        ExecStart = ''
+          ${webhook}/bin/webhook -hooks ${configFile} ${lib.concatStringsSep " " cfg.extraOptions}
+        '';
+      };
+    };
+
+    users = {
+      users = mkIf (cfg.user == null) {
+        webhook.isSystemUser = true;
+        inherit group;
```

sorry about this one, made a mistake in the earlier review comment. `group` is an attribute of the user `webhook`, so it really should've been this in the first suggestion:
```suggestion
      users.webhook = mkIf (cfg.user == null) {
        isSystemUser = true;
        inherit group;
```
the point is to set the group of the created user correctly when it is created, since system users don't get a default group.

---

Is there a reason to not follow the standard convention of providing a default value of "webhook" and provisioning the user based on that? Almost every other module does this.

---

That's what I had first, but @pennae advised me to do it with `default = null`

---

@aanderse kinda told them not to (https://github.com/NixOS/nixpkgs/pull/152333#discussion_r775498817) from having seen many a module do the opposite too, but on second thought it *is* much better to always have strings in these attributes for interpolation into other parts of a config :thinking: perhaps those other modules that use `null` for a default should be changed.

---

Which modules use `null`?

---

at least amuled and quassel, i'm pretty sure there's a few more but can't remember off the top of my head.

---

We should look at standardizing those...

---

I've changed it back to `webhook:webhook`.